In [6]:
# Librerias

import pandas as pd
import numpy as np
import plotly.graph_objects as go
import plotly.express as px
import statsmodels.api as sm
from plotly.subplots import make_subplots

In [7]:
# Jaime

#df_calendar = pd.read_csv("/Users/jaime/Documents/ICAI/Quinto/Desarrollo Apps de Visualización/Trabajo/calendar.csv")
#df_listings = pd.read_csv("/Users/jaime/Documents/ICAI/Quinto/Desarrollo Apps de Visualización/Trabajo/listings.csv")
#df_neighbourhoods = pd.read_csv("/Users/jaime/Documents/ICAI/Quinto/Desarrollo Apps de Visualización/Trabajo/neighbourhoods.csv")
#df_reviews = pd.read_csv("/Users/jaime/Documents/ICAI/Quinto/Desarrollo Apps de Visualización/Trabajo/reviews.csv")
#df_reviews_det = pd.read_csv("/Users/jaime/Documents/ICAI/Quinto/Desarrollo Apps de Visualización/Trabajo/reviews_detailed.csv")
#df_listings_det = pd.read_csv("/Users/jaime/Documents/ICAI/Quinto/Desarrollo Apps de Visualización/Trabajo/listings_detailed.csv")

In [8]:
# Diego

df_calendar = pd.read_csv("/Users/diegoma/kaggle/calendar.csv")
df_listings = pd.read_csv("/Users/diegoma/kaggle/listings.csv")
df_neighbourhoods = pd.read_csv("/Users/diegoma/kaggle/neighbourhoods.csv")
df_reviews = pd.read_csv("/Users/diegoma/kaggle/reviews.csv")
df_reviews_det = pd.read_csv("/Users/diegoma/kaggle/reviews_detailed.csv")
df_listings_det = pd.read_csv("/Users/diegoma/kaggle/listings_detailed.csv")

In [9]:
#df_listings_det = pd.read_csv("/Users/diegoma/kaggle/listings_detailed.csv")

In [10]:
df_calendar.head()

listing_id        date available   price adjusted_price  minimum_nights  \
0      167183  2021-04-15         f  $45.00         $45.00             1.0   
1        6369  2021-04-15         t  $60.00         $60.00             1.0   
2        6369  2021-04-16         t  $60.00         $60.00             1.0   
3        6369  2021-04-17         t  $60.00         $60.00             1.0   
4        6369  2021-04-18         t  $60.00         $60.00             1.0   

   maximum_nights  
0             5.0  
1          1125.0  
2          1125.0  
3          1125.0  
4          1125.0

In [11]:
df_calendar.isna().sum()

listing_id           0
date                 0
available            0
price             1105
adjusted_price    1105
minimum_nights    1050
maximum_nights    1050
dtype: int64

In [12]:
df_listings.head()

id                                      name    host_id host_name  \
0   6369  Rooftop terrace room ,  ensuite bathroom      13660     Simon   
1  21853                      Bright and airy room      83531     Abdel   
2  23001         Apartmento Arganzuela- Madrid Rio      82175     Jesus   
3  24805                    Gran Via Studio Madrid  346366726         A   
4  26825        Single Room whith private Bathroom     114340  Agustina   

  neighbourhood_group   neighbourhood  latitude  longitude        room_type  \
0           Chamartín  Hispanoamérica  40.45724   -3.67688     Private room   
1              Latina        Cármenes  40.40381   -3.74130     Private room   
2          Arganzuela         Legazpi  40.38840   -3.69511  Entire home/apt   
3              Centro     Universidad  40.42183   -3.70529  Entire home/apt   
4          Arganzuela         Legazpi  40.38975   -3.69018     Private room   

   price  minimum_nights  number_of_reviews last_review  reviews_per_month  \
0     60               1                 78  2020-09-20               0.58   
1     31               4                 33  2018-07-15               0.42   
2     50              15                  0         NaN                NaN   
3     92               5                 10  2020-03-01               0.13   
4     26               2                149  2020-03-12               1.12   

   calculated_host_listings_count  availability_365  
0                               1               180  
1                               2               364  
2                               7                 1  
3                               1                72  
4                               1               365

In [13]:
df_listings.isna().sum()

id                                   0
name                                 3
host_id                              0
host_name                          527
neighbourhood_group                  0
neighbourhood                        0
latitude                             0
longitude                            0
room_type                            0
price                                0
minimum_nights                       0
number_of_reviews                    0
last_review                       5637
reviews_per_month                 5637
calculated_host_listings_count       0
availability_365                     0
dtype: int64

In [14]:
df_reviews_det.head()

listing_id     id        date  reviewer_id reviewer_name  \
0        6369  29428  2010-03-14        84790         Nancy   
1        6369  31018  2010-03-23        84338         David   
2        6369  34694  2010-04-10        98655        Marion   
3        6369  37146  2010-04-21       109871          Kurt   
4        6369  38168  2010-04-26        98901        Dennis   

                                            comments  
0  Simon and Arturo have the ultimate location in...  
1  Myself and Kristy originally planned on stayin...  
2  We had a great time at Arturo and Simon's ! A ...  
3  I very much enjoyed the stay.  \r\nIt's a wond...  
4  Arturo and Simon are polite and friendly hosts...

In [15]:
df_reviews_det.isna().sum()

listing_id         0
id                 0
date               0
reviewer_id        0
reviewer_name      1
comments         352
dtype: int64

In [16]:
df_listings_det.head()

id                         listing_url       scrape_id last_scraped  \
0   6369   https://www.airbnb.com/rooms/6369  20210413152844   2021-04-15   
1  21853  https://www.airbnb.com/rooms/21853  20210413152844   2021-04-16   
2  23001  https://www.airbnb.com/rooms/23001  20210413152844   2021-04-16   
3  24805  https://www.airbnb.com/rooms/24805  20210413152844   2021-04-15   
4  26825  https://www.airbnb.com/rooms/26825  20210413152844   2021-04-15   

                                       name  \
0  Rooftop terrace room ,  ensuite bathroom   
1                      Bright and airy room   
2         Apartmento Arganzuela- Madrid Rio   
3                    Gran Via Studio Madrid   
4        Single Room whith private Bathroom   

                                         description  \
0  Excellent connection with the AIRPORT and EXHI...   
1  We have a quiet and sunny room with a good vie...   
2  Apartamento de tres dormitorios dobles, gran s...   
3  Studio located 50 meters from Gran Via, next t...   
4  Nice and cozy roon for one person with a priva...   

                               neighborhood_overview  \
0                                                NaN   
1  We live in a leafy neighbourhood with plenty o...   
2  Barrio Arganzuela, junto a Madrid Rio, zonas c...   
3  The area is next to the Gran Via, so people li...   
4  Es un barrio muy tranquilo, en una zona de Mad...   

                                         picture_url    host_id  \
0  https://a0.muscache.com/pictures/683224/4cc318...      13660   
1  https://a0.muscache.com/pictures/68483181/87bc...      83531   
2  https://a0.muscache.com/pictures/58e6a770-509c...      82175   
3  https://a0.muscache.com/pictures/miso/Hosting-...  346366726   
4  https://a0.muscache.com/pictures/149358/218d5b...     114340   

                                      host_url  ...  \
0      https://www.airbnb.com/users/show/13660  ...   
1      https://www.airbnb.com/users/show/83531  ...   
2      https://www.airbnb.com/users/show/82175  ...   
3  https://www.airbnb.com/users/show/346366726  ...   
4     https://www.airbnb.com/users/show/114340  ...   

  review_scores_communication review_scores_location review_scores_value  \
0                        10.0                   10.0                10.0   
1                        10.0                    8.0                 9.0   
2                         NaN                    NaN                 NaN   
3                        10.0                   10.0                10.0   
4                         9.0                    9.0                 9.0   

  license instant_bookable calculated_host_listings_count  \
0     NaN                f                              1   
1     NaN                f                              2   
2     NaN                f                              7   
3     NaN                t                              1   
4     NaN                f                              1   

  calculated_host_listings_count_entire_homes  \
0                                           0   
1                                           0   
2                                           6   
3                                           1   
4                                           0   

  calculated_host_listings_count_private_rooms  \
0                                            1   
1                                            2   
2                                            1   
3                                            0   
4                                            1   

  calculated_host_listings_count_shared_rooms reviews_per_month  
0                                           0              0.58  
1                                           0              0.42  
2                                           0               NaN  
3                                           0              0.13  
4                                           0              1.12  

[5 rows x 74 columns]

In [17]:
df_listings_det.isna().sum()

id                                                 0
listing_url                                        0
scrape_id                                          0
last_scraped                                       0
name                                               3
                                                ... 
calculated_host_listings_count                     0
calculated_host_listings_count_entire_homes        0
calculated_host_listings_count_private_rooms       0
calculated_host_listings_count_shared_rooms        0
reviews_per_month                               5637
Length: 74, dtype: int64

Comenzaremos explorando brevemente el precio en función de diferentes variables. Primero tendrá sentido analizar el efecto del barrio en el precio


In [18]:
df_neighbourhoods

neighbourhood_group  neighbourhood
0            Arganzuela        Acacias
1            Arganzuela         Atocha
2            Arganzuela        Chopera
3            Arganzuela       Delicias
4            Arganzuela       Imperial
..                  ...            ...
123          Villaverde       Butarque
124          Villaverde    Los Angeles
125          Villaverde    Los Rosales
126          Villaverde     San Andrés
127          Villaverde  San Cristobal

[128 rows x 2 columns]

In [19]:
df_neighbourhoods["neighbourhood_group"].unique()

array(['Arganzuela', 'Barajas', 'Carabanchel', 'Centro', 'Chamartín',
       'Chamberí', 'Ciudad Lineal', 'Fuencarral - El Pardo', 'Hortaleza',
       'Latina', 'Moncloa - Aravaca', 'Moratalaz', 'Puente de Vallecas',
       'Retiro', 'Salamanca', 'San Blas - Canillejas', 'Tetuán', 'Usera',
       'Vicálvaro', 'Villa de Vallecas', 'Villaverde'], dtype=object)

In [20]:
len(df_neighbourhoods["neighbourhood_group"].unique())

21

21 diferentes barrios globales o distritos

In [21]:
df_neighbourhoods["neighbourhood"].unique()

array(['Acacias', 'Atocha', 'Chopera', 'Delicias', 'Imperial', 'Legazpi',
       'Palos de Moguer', 'Aeropuerto', 'Alameda de Osuna',
       'Casco Histórico de Barajas', 'Corralejos', 'Timón', 'Abrantes',
       'Buenavista', 'Comillas', 'Opañel', 'Puerta Bonita', 'San Isidro',
       'Vista Alegre', 'Cortes', 'Embajadores', 'Justicia', 'Palacio',
       'Sol', 'Universidad', 'Castilla', 'Ciudad Jardín', 'El Viso',
       'Hispanoamérica', 'Nueva España', 'Prosperidad', 'Almagro',
       'Arapiles', 'Gaztambide', 'Rios Rosas', 'Trafalgar',
       'Vallehermoso', 'Atalaya', 'Colina', 'Concepción', 'Costillares',
       'Pueblo Nuevo', 'Quintana', 'San Juan Bautista', 'San Pascual',
       'Ventas', 'El Goloso', 'El Pardo', 'Fuentelareina', 'La Paz',
       'Mirasierra', 'Peñagrande', 'Pilar', 'Valverde',
       'Apostol Santiago', 'Canillas', 'Palomas', 'Pinar del Rey',
       'Piovera', 'Valdefuentes', 'Aguilas', 'Aluche', 'Campamento',
       'Cármenes', 'Cuatro Vientos', 'Lucero', '

In [22]:
len(df_neighbourhoods["neighbourhood"].unique())

128

128 barrios en total

Es de interés, en primer lugar, ver la distribución de los precios

In [23]:
data = [
    go.Histogram(
        x = df_listings['price'],
        opacity=0.6,
        name = "Precio"
    )
]

layout = go.Layout(title = "Distribución del precio", xaxis_title = "Precio", yaxis_title = "Frecuencia",
                   barmode = "overlay")

fig = go.Figure(data = data, layout = layout)

fig.show()

In [24]:
df_listings['price'][df_listings['price'] > 500].count()

610

La cantidad de outliers es elevada, existiendo 610 apartamentos ofrecidos por un precio por noche superior a 500€. Se deberá considerar como actuar frente a dicha observación. A continuación, para ver realmente la distribución del grueso de publicaciones, filtramos los precios menores a 300€:

In [25]:
data = [
    go.Histogram(
        x = df_listings['price'][df_listings['price'] < 300],
        opacity=0.6,
        name = "Precio"
    )
]

layout = go.Layout(title = "Distribución del precio", xaxis_title = "Precio", yaxis_title = "Frecuencia",
                   barmode = "overlay")

fig = go.Figure(data = data, layout = layout)

fig.show()

A continuación, analizaremos el impacto del barrio. Primero de todo, cual es la distribución de acuerdo al barrio:

In [26]:
count_neigh = len(df_neighbourhoods["neighbourhood_group"].unique())
unique_neigh = df_neighbourhoods["neighbourhood_group"].unique()
neigh_count = {}

for i in range(count_neigh):
  neigh_count[unique_neigh[i]] = df_listings["neighbourhood_group"][df_listings["neighbourhood_group"] == unique_neigh[i]].count()

In [37]:
ordered_keys = []
ordered_values = []

for w in sorted(neigh_count, key=neigh_count.get):
    ordered_keys.append(w)
    ordered_values.append(neigh_count[w])

fig = go.Figure(go.Bar(
            x=ordered_values,
            y=ordered_keys,
            orientation='h', 
            marker_color = 'firebrick'))

fig.update_layout(title_text = "Numero de publicaciones por barrio", xaxis_title = "Numero de publicaciones", yaxis_title = "Barrio")

fig.show()

Existe un muy claro dominio de ofertas de publicaciones en el barrio "Centro"

In [28]:
price_neigh = {}
for i in range(count_neigh):
  price_neigh[unique_neigh[i]] = round(df_listings[df_listings["neighbourhood_group"] == unique_neigh[i]]["price"].mean(),3)

In [29]:
price_neigh

{'Arganzuela': 140.022,
 'Barajas': 119.849,
 'Carabanchel': 85.325,
 'Centro': 127.052,
 'Chamartín': 112.745,
 'Chamberí': 115.53,
 'Ciudad Lineal': 105.391,
 'Fuencarral - El Pardo': 117.908,
 'Hortaleza': 138.646,
 'Latina': 137.406,
 'Moncloa - Aravaca': 114.827,
 'Moratalaz': 91.737,
 'Puente de Vallecas': 78.246,
 'Retiro': 124.565,
 'Salamanca': 153.239,
 'San Blas - Canillejas': 343.016,
 'Tetuán': 119.734,
 'Usera': 93.554,
 'Vicálvaro': 138.903,
 'Villa de Vallecas': 75.735,
 'Villaverde': 180.068}

In [30]:
# Cómo es la distribución de precios de acuerdo al barrio? 

data = [
    go.Bar(
        x = list(price_neigh.keys()),
        y = list(price_neigh.values()),
        name = "Average price per neighbourhood",
        marker_color = ["aliceblue","antiquewhite","aqua","beige","bisque","black", "blueviolet",
                        "cornsilk", "darkkhaki", "darkgrey", "darkgreen", "firebrick", "gainsboro",
                        "ivory", "crimson", "darkslategray", "indigo", "mediumaquamarine", "olivedrab", "peachpuff","mediumspringgreen"],
    )
]

layout = go.Layout(title = "Distribución de precios en función del distrito", xaxis_title = "Distrito", 
                   yaxis_title = "Precio medio",
                  )

fig = go.Figure(data = data, layout = layout)

fig.show()

San Blas - Canillejas parece que pueda ser un outlier

In [31]:
data = [
    go.Box(
        y = df_listings[df_listings["neighbourhood_group"] == "San Blas - Canillejas"]["price"],
        marker_color = "steelblue",
        name = "Precio",
        boxmean=True
    )
]

layout = go.Layout(title = "Distribución de las precios en San Blas", yaxis_title = "Precio")

fig = go.Figure(data = data, layout = layout)

fig.show()

Como podemos comprobar, existe un apartamento con un precio mucho mayor al resto, lo que hace que incremente la media del distrito. Por ello, comprobaremos las medianas de cada distrito:

In [32]:
median_neigh = {}
for i in range(count_neigh):
  median_neigh[unique_neigh[i]] = round(df_listings[df_listings["neighbourhood_group"] == unique_neigh[i]]["price"].median(),3)

In [33]:
# Cómo es la distribución de precios de acuerdo al barrio? 

data = [
    go.Bar(
        x = list(median_neigh.keys()),
        y = list(median_neigh.values()),
        name = "Median price per neighbourhood",
        marker_color = ["aliceblue","antiquewhite","aqua","beige","bisque","black", "blueviolet",
                        "cornsilk", "darkkhaki", "darkgrey", "darkgreen", "firebrick", "gainsboro",
                        "ivory", "crimson", "darkslategray", "indigo", "mediumaquamarine", "olivedrab", "peachpuff","mediumspringgreen"],
    )
]

layout = go.Layout(title = "Distribución de precios en función del distrito", xaxis_title = "Distrito", 
                   yaxis_title = "Precio mediano por noche",
                  )

fig = go.Figure(data = data, layout = layout)

fig.show()

Pese a que San Blas sigue siendo uno de los distritos de mayor precio, la distribucion entre distritos es de mayor uniformidad

A continuación, mostraremos un mapa de calor con el precio de los diferentes apartamentos:

In [34]:
fig = px.density_mapbox(df_listings, lat=df_listings['latitude'], lon=df_listings['longitude'], z=df_listings['price'], radius=10,
                        center=dict(lat=40.43, lon=-3.68), zoom=11.5,
                        mapbox_style="carto-positron", labels = {"price": "Price", "latitude": "Latitude", "longitude":"Longitude", "neighbourhood":"Neighbourhood"},
                        hover_data = ["price", "latitude", "longitude", df_listings['neighbourhood']],
                        width = 1200, height = 800)
fig.show()

Otra variable a tener en cuenta será el tipo de habitación. Por ello, estudiaremos el precio medio de acuerdo al tipo de habitación. Primero de todo, cuantos tipos de habitación hay:

In [35]:
level_count = pd.DataFrame(df_listings["room_type"].value_counts()).reset_index().rename(columns = {"index":"room_type", "room_type": "count"})
level_count = level_count.sort_values("room_type").reset_index(drop = True)

fig = px.pie(level_count, values="count", names="room_type", color = "room_type",
            color_discrete_map = {
                "Private room": "lightblue",
                "Entire home/apt": "mediumseagreen",
                "Shared room": "gold",
                "Hotel room": "darkorange",                
            })
fig.show()

Como era de esperar, la gran mayoría de habitaciones son privadas o apartamentos enteros. Ahora veremos la relación de precios:

In [36]:
median_room = {}
mean_room = {}
unique_room = df_listings["room_type"].unique()
count_room = len(unique_room)

for i in range(count_room):
  median_room[unique_room[i]] = round(df_listings[df_listings["room_type"] == unique_room[i]]["price"].median(),2)
  mean_room[unique_room[i]] = round(df_listings[df_listings["room_type"] == unique_room[i]]["price"].mean(),2)

fig = make_subplots(rows = 1,
                    cols = 2,
                    subplot_titles=("Precio mediano por tipo de habitación", "Precio medio por tipo de habitación"))
fig.add_trace(
    go.Bar(
        x = list(median_room.keys()),
        y = list(median_room.values()),
        name = "Precio mediano por tipo de habitación",
        marker_color = ["darkgreen", "firebrick",
                        "darkslategray", "indigo"],
    ),
    row = 1, col = 1    
)

fig.add_trace(
    go.Bar(
        x = list(mean_room.keys()),
        y = list(mean_room.values()),
        name = "Precio medio por tipo de habitación",
        marker_color = ["darkgreen", "firebrick",
                        "darkslategray", "indigo"],
    ),
    row = 1, col = 2    
)

fig.update_layout(width = 1500, height = 600, title = "Precios según tipo de habitación", showlegend = False)

fig.show()

Como podemos ver, existe una diferencia sustancial, tanto en terminos de media como de mediana, en función del tipo de habitación, por lo que será una variable importante a la hora de desarrollar el modelo